stripe/src/main/java/com/stripe/android/net/PollingNetworkHandler.java

```diff
@@ -0,0 +1,188 @@
+    private static final int EXPIRED = -2;
+
+    private final String mClientSecret;
+    private final String mPublishableKey;
+    private final String mSourceId;
+
+    private final long mTimeoutMs;
+    private final boolean mIsInSingleThreadMode;
+    @NonNull private final SourceRetriever mSourceRetriever;
+
+    private Handler mNetworkHandler;
+    private Handler mUiHandler;
+
+    private final Runnable pollRunnable = new Runnable() {
+        @Override
+        public void run() {
+            Message message = null;
+            try {
+                Source source = mSourceRetriever.retrieveSource(
+                        mSourceId,
+                        mClientSecret,
+                        mPublishableKey);
+                if (source.getRedirect() != null) {
+                    switch (source.getRedirect().getStatus()) {
+                        case SourceRedirect.SUCCEEDED:
+                            message = mUiHandler.obtainMessage(SUCCESS);
+                            break;
+                        case SourceRedirect.PENDING:
+                            message = mUiHandler.obtainMessage(PENDING);
+                            break;
+                        case SourceRedirect.FAILED:
+                            message = mUiHandler.obtainMessage(FAILURE);
+                            break;
+                    }
+                }
+            } catch (StripeException stripeEx) {
+                message = mUiHandler.obtainMessage(ERROR, stripeEx);
+            } finally {
+                if (message != null) {
+                    mUiHandler.sendMessage(message);
+                }
+            }
+        }
+    };
+
+    PollingNetworkHandler(@NonNull final String sourceId,
+                          @NonNull final String clientSecret,
+                          @NonNull final String publishableKey,
+                          @NonNull final PollingResponseHandler callback,
+                          @Nullable Integer timeOutMs,
+                          @Nullable SourceRetriever sourceRetriever) {
+        mSourceId = sourceId;
+        mClientSecret = clientSecret;
+        mPublishableKey = publishableKey;
+        mIsInSingleThreadMode = sourceRetriever != null;
+        mSourceRetriever = sourceRetriever == null
+                ? new SourceRetriever() {
+                    @Override
+                    public Source retrieveSource(
+                            @NonNull String sourceId,
+                            @NonNull String clientSecret,
+                            @NonNull String publishableKey) throws StripeException {
+                        return StripeApiHandler.retrieveSource(
+                                sourceId,
+                                clientSecret,
+                                publishableKey);
+                    }
+                }
+                : sourceRetriever;
+
+        mTimeoutMs = timeOutMs == null ? DEFAULT_TIMEOUT_MS : timeOutMs.longValue();
```

Comment:
It's probably wise to also cap the timeout – we want to make sure apps aren't sitting around polling an API endpoint for a really long time, as that could create a lot of load. We use 5 minutes on iOS.

https://github.com/stripe/stripe-ios/blob/master/Stripe/STPSourcePoller.m#L19